MT3-380 #to-do Manager review page (part 1)

Gweaton opened the pull request:
This is the first half of the manager review page flow. It adds a manager version of the Review page with approve & decline buttons that lead onto a confirmation page, and make the corresponding requests to the MaT API to update this data.

To come in this PR:
- Update Readme with instructions on how to test the process

To come in part 2:
- API endpoints for displaying data from Outsystems at top of Review page
- Readonly-manager review page with additional notes section

Gweaton committed:
Extract clearing of Session Storage

We now only want to clear the session storage after we've transferred
the process to the manager

erbridge commented on components/ManagerSubmitButton.tsx:
```diff
@@ -0,0 +1,54 @@
+  return (
+    <Button
+      preventDoubleClick
+      className={
+        buttonText === "Decline"
+          ? "lbh-button--warning govuk-button--warning"
+          : "submit-button"
+      }
```
There's a package we have installed to simplify this `classnames` (if we don't, I'm surprized, and it's worth installing).

Also, the `"submit-button"` class was for styling that we don't have here (but maybe should?). If you're seeing styling, it's likely a side effect of using global styles on `makeSubmit`, and we should use a different class name.

```suggestion
      className={
        classNames({ "lbh-button--warning govuk-button--warning": status === "decline" })
      }
```

erbridge commented on components/ManagerSubmitButton.tsx:
```diff
@@ -0,0 +1,54 @@
+import urlsForRouter from "helpers/urlsForRouter";
```
```suggestion
import urlsForRouter from "../helpers/urlsForRouter";
```

erbridge commented on components/ManagerSubmitButton.tsx:
```diff
@@ -0,0 +1,54 @@
+import { NextRouter, useRouter } from "next/router";
+import PropTypes from "prop-types";
+import React from "react";
+import PageSlugs, { urlObjectForSlug } from "steps/PageSlugs";
```
```suggestion
import PageSlugs, { urlObjectForSlug } from "../steps/PageSlugs";
```

erbridge commented on components/ManagerSubmitButton.tsx:
```diff
@@ -0,0 +1,54 @@
+
+interface Props {
+  onSubmit(router: NextRouter): Promise<void>;
+  buttonText: string;
```
Generally, prefer `children` when specifying a prop that's used as the child of the main element of the component. This could be any React node.

Gweaton replied:
Good to know - would we still want to do this if we're passing the "status" through as an enum?

erbridge replied:
No. This was a note for future reference. With status, it's definitely a normal property.

erbridge commented on components/ManagerSubmitButton.tsx:
```diff
@@ -0,0 +1,54 @@
+
+        try {
+          await onSubmit(router);
+          await router.push(href, as);
```
Let's pull the router push outside the `try..catch` to be consistent with the behaviour of `makeSubmit`. Might even be worth refactoring that's `onClick` method to be used here. I'd expect them to look pretty similar.

Gweaton replied:
They do look similar, do you mean extract the `onClick` into more of a shared function, or use `makeSubmit` itself? Obviously the former is a bit less work, although I'm unsure where it would go - a helper maybe?

erbridge replied:
I mean pull out the common logic into something more easily shared.

erbridge commented on components/ManagerSubmitButton.tsx:
```diff
@@ -0,0 +1,54 @@
+
+  const { href, as } = urlsForRouter(router, {
+    ...urlObjectForSlug(router, PageSlugs.Confirmed),
+    query: { status: status },
```
I like this approach!

erbridge commented on components/ManagerSubmitButton.tsx:
```diff
@@ -0,0 +1,54 @@
+  const status = buttonText === "Approve" ? "approved" : "declined";
+
+  const { href, as } = urlsForRouter(router, {
+    ...urlObjectForSlug(router, PageSlugs.Confirmed),
```
Should this go to a new `ManagerSubmit` page? We don't want to persist any actions if they're offline (which is technically possible on this page). That's what the `Submit` and `Confirmed` pair of pages are for in the normal process, and I think we need the same here.

erbridge replied:
On second thought, I think this is actually right, as long as the manager can't make any changes to the process itself. We should probably adjust the loading page text to assume that a manager is online, though, even if we handle them being offline, as we expect them to be reviewing from their desktops.

erbridge reviewed (commented):
Review of the manager submit button. I'll take a look at the rest tomorrow.

erbridge commented on .env.example:
```diff
@@ -23,3 +23,4 @@ TEST_PROCESS_REF=
 TEST_PROCESS_API_JWT=
 TEST_MAT_API_JWT=
 TEST_MAT_API_DATA=
+TEST_PROCESS_STAGE='0' # or 1
```
```suggestion
TEST_PROCESS_STAGE='0' # or 1

```

erbridge commented on helpers/approveProcess.ts:
```diff
@@ -0,0 +1,53 @@
+import getProcessRef from "./getProcessRef";
+import isClient from "./isClient";
+
+const approveProcess = async (router: NextRouter): Promise<void> => {
```
A bunch of this code is common to both approval and declining. Let's pull it out to be shared.

erbridge commented on pages/thc/[processRef]/confirmed.tsx:
```diff
@@ -6,36 +6,52 @@ import {
-          Go to diversity monitoring form
-        </Button>
-      </Paragraph>
+      {!isManager ? (
```
```suggestion
      {isManager || (
```

erbridge commented on pages/thc/[processRef]/review.tsx:
```diff
@@ -1,4 +1,6 @@
+import ManagerSubmitButton from "components/ManagerSubmitButton";
```
```suggestion
import ManagerSubmitButton from "../../../components/ManagerSubmitButton";
```

erbridge commented on pages/thc/[processRef]/review.tsx:
```diff
@@ -1,4 +1,6 @@
+import ManagerSubmitButton from "components/ManagerSubmitButton";
 import formatDate from "date-fns/format";
+import declineProcess from "helpers/declineProcess";
```
```suggestion
import declineProcess from "../../../helpers/declineProcess";
```

erbridge commented on pages/thc/[processRef]/review.tsx:
```diff
@@ -32,6 +36,7 @@ const ReviewPage: NextPage = () => {
   const router = useRouter();
 
   const processRef = getProcessRef(router);
+  const isManager = getProcessStage(router) === "1";
```
Rather than using the magic number everywhere, can we pull this comparison out into a method of its own?

erbridge commented on pages/thc/[processRef]/review.tsx:
```diff
@@ -162,46 +167,59 @@ const ReviewPage: NextPage = () => {
           setSignatures((sigs) => ({ ...sigs, [selectedTenantId]: value }));
         }}
+        readOnly={isManager}
       />
```
Note that `Signature` isn't actually populated from the database (intentionally - see the comment by the `useState` call). We probably want to show images instead of the component for the manager.

Gweaton replied:
This actually crept in, I removed it in this PR because I'm going to handle it in the next one, but this is good to know for the future.

erbridge commented on pages/thc/[processRef]/review.tsx:
```diff
@@ -162,46 +167,59 @@ const ReviewPage: NextPage = () => {
+          <ManagerSubmitButton onSubmit={approveProcess} buttonText="Approve" />
+          <ManagerSubmitButton onSubmit={declineProcess} buttonText="Decline" />
+          <style jsx>{`
+            :global(.submit-button:not(:last-child)) {
```
This classname should probably change to avoid any accidental style clashes.

Gweaton committed:
Store ProcessStage in session storage

To allow us to determine whether a process has been assigned to an
officer (set to '0') or a manager (set to '1'). There are also statuses
for when a check has been approved ('2') or declined ('3') by a manager.

Gweaton committed:
Add Transfer API endpoint

This is called to transfer a process from Officer to Manager and
vice-versa (when approving/declining a process)

Gweaton committed:
Transfer process to manager on submit

Gweaton committed:
Extract fetching MaT API data from Session Storage

Gweaton committed:
Add /appraise endpoint and frontend query methods

These will be called when a manager approves/declines a process

Gweaton committed:
Add Classnames package

This makes it easier to conditionally add classNames to elements

Gweaton committed:
Add Manager page flow

If we get a processStage of 1 back from the MaT api, we treat it as the
manager's process.

Managers go straight from the loading screen to the Manager Review page

Gweaton committed:
Add Manager Review & Confirmation pages

Gweaton commented on server/api.js:
```diff
@@ -204,6 +204,142 @@ router.post("/v1/processes/:ref/images", (req, res) => {
+    processStage: processStage,
+    serviceRequest: {
+      id: serviceRequestId,
+      title: "Tenancy Management",
```
This is just a hardcoded value according to https://hackney.atlassian.net/browse/MT3-381 - @LBHTKarki if you have any further information on this, that would be really helpful!

Gweaton committed:
Extract the logic for submitting a step into a helper

We want to share it between the step submission button and the manager
submission button, as they need to do basically the same thing.

Gweaton committed:
Encode query params

We've seen an issue where forward slashes are being double encoded, so
this should hopefully resolve the issue

erbridge committed:
Make feature tests disable process updates to the MaT API

Real set up of a process involves assigning it to an officer, which will
require some work to make sure we're doing cleanup correctly after the
tests have run. This is a quick fix that unblocks the rest of the
manager review work.

Gweaton commented on helpers/ProcessStage.ts:
```diff
@@ -0,0 +1,6 @@
+export enum ProcessStage {
+  InProgress = "0",
+  InReview = "1",
```
👍

Gweaton commented:
I can't approve as I'm technically the author of this, but looks good to me!
